### PR TITLE
Add checkpoint artifact path prefix to MLflow logger

### LINKED ISSUE #20540: Add new `checkpoint_artifact_path_prefix` argument `MLFlowLogger` to customize artifact path for model checkpoints
### Description & Motivation

See https://github.com/Lightning-AI/pytorch-lightning/discussions/19394#discussioncomment-11771861
PyTorch Lightning's `MLFlowLogger` changed its behavior and there are things that are inconvenient about the current implementation.

### Pitch

I have opened a PR to fix this as per above.

### Alternatives

_No response_

### Additional context

_No response_

cc @lantiga @borda

### PR DESCRIPTION
**Pull request by benglewis**
## What does this PR do?

Introduces a new `checkpoint_artifact_path_prefix` argument to `MLFlowLogger` which customizes the prefix for the model artifacts.

Fixes #20540

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

Add a new `checkpoint_artifact_path_prefix` parameter to the MLflow logger.

* Modify `src/lightning/pytorch/loggers/mlflow.py` to include the new parameter in the `MLFlowLogger` class constructor and use it in the `after_save_checkpoint` method.
* Update the documentation in `docs/source-pytorch/visualize/loggers.rst` to include the new `checkpoint_artifact_path_prefix` parameter.
* Add a new test in `tests/tests_pytorch/loggers/test_mlflow.py` to verify the functionality of the `checkpoint_artifact_path_prefix` parameter and ensure it is used in the artifact path.

---

For more details, open the [Copilot Workspace session](https://copilot-workspace.githubnext.com/Lightning-AI/pytorch-lightning/pull/20538?shareId=4e707115-0a9c-49c1-a9d2-9c051d0c140e).

----
📚 Documentation preview 📚: https://pytorch-lightning--20538.org.readthedocs.build/en/20538/

### CODE REVIEW

File: docs/source-pytorch/visualize/loggers.rst [hunk 1]
```diff
@@ -54,3 +54,37 @@ Track and Visualize Experiments
...
+MLflow Logger
+-------------
+
+The MLflow logger in PyTorch Lightning now includes a `checkpoint_artifact_path_prefix` parameter. This parameter allows you to prefix the checkpoint artifact's path when logging checkpoints as artifacts.
+
+Example usage:
+
+.. code-block:: python
+
+    from lightning.pytorch import Trainer
```
Comment (lantiga) [inline docs/source-pytorch/visualize/loggers.rst:69]:
```suggestion
    import lightning as L
```

File: docs/source-pytorch/visualize/loggers.rst [hunk 2]
```diff
@@ -54,3 +54,37 @@ Track and Visualize Experiments
...
+    from lightning.pytorch.loggers import MLFlowLogger
+
+    mlf_logger = MLFlowLogger(
+        experiment_name="lightning_logs",
+        tracking_uri="file:./ml-runs",
+        checkpoint_artifact_path_prefix="my_prefix"
+    )
+    trainer = Trainer(logger=mlf_logger)
```
Comment (lantiga) [inline docs/source-pytorch/visualize/loggers.rst:77]:
```suggestion
    trainer = L.Trainer(logger=mlf_logger)
```

File: docs/source-pytorch/visualize/loggers.rst [hunk 3]
```diff
@@ -54,3 +54,37 @@ Track and Visualize Experiments
...
+
+    # Your LightningModule definition
+    class LitModel(LightningModule):
```
Comment (lantiga) [inline docs/source-pytorch/visualize/loggers.rst:80]:
```suggestion
    class LitModel(L.LightningModule):
```

File: docs/source-pytorch/visualize/loggers.rst [hunk 4]
```diff
@@ -54,3 +54,37 @@ Track and Visualize Experiments
 
         </div>
     </div>
+
+.. _mlflow_logger:
+
+MLflow Logger
+-------------
+
+The MLflow logger in PyTorch Lightning now includes a `checkpoint_artifact_path_prefix` parameter. This parameter allows you to prefix the checkpoint artifact's path when logging checkpoints as artifacts.
```
Comment (lantiga) [inline docs/source-pytorch/visualize/loggers.rst:63]: I would simplify the name of the argument to be `checkpoint_path_prefix`. The extra `artifact` doesn't add much imo and simpler is better, wdyt?

**Review by lantiga — COMMENTED** [review]
Thanks a lot of the PR, I left a couple of comments but we're good to go.

Tests need fixing, can you take that @benglewis?

**Comment by lantiga** [comment]
thank you @benglewis , happy to merge once tests are fixed

**Comment by benglewis** [comment]
@lantiga I have commit a change that fixes the test when I ran it with `pytest` :)
This will be a great improvement for us in the next `pytorch-lightning` release

**Comment by benglewis** [comment]
@lantiga Do you need any additional changes here? 🙏

File: src/lightning/pytorch/CHANGELOG.md [hunk 5]
```diff
@@ -4,6 +4,8 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
+- Added a new `checkpoint_artifact_path_prefix` parameter to the MLflow logger which can control the path to where the MLflow artifacts for the model checkpoints are stored.
```
Comment (joncarter1) [inline src/lightning/pytorch/CHANGELOG.md:7]: @benglewis doc and example are stale but otherwise looks great, thanks for fixing this!

I'd just complained on the original PR (https://github.com/Lightning-AI/pytorch-lightning/pull/20325#issuecomment-2637363515) that caused this as it silently broke my code innocuously too!

Comment (benglewis) [reply src/lightning/pytorch/CHANGELOG.md:7]: Thank you for the reminder to update the docs

**Comment by benglewis** [comment]
@lantiga I just updated the stale documentation as per @joncarter1 's comment. It should be ready for merge now 🙏

**Comment by benglewis** [comment]
@lantiga Do I need to do something to runt he `check/schema` and `required-jobs` checks?

**Comment by benglewis** [comment]
@Borda @ethanwharris @tchaton @justusschock @lantiga Is there anything in particular holding this up?

**Comment by benglewis** [comment]
@Borda Is there anything that I need to change to get this PR merged?

**Comment by benglewis** [comment]
@Borda @ethanwharris @tchaton @justusschock @lantiga 
I don't like "nagging" and such on GitHub PRs, and I understand and image that you are all extremely busy at PyTorch Lightning, I was just wondering what is preventing this PR from being merged :/

**Comment by benglewis** [comment]
@Borda I see that you have reverted this PR and I see why:
It seems like I inadvertently introduced [another bug](https://github.com/mlflow/mlflow/issues/15111), which I only noticed now since I am also affected :|
Would you accept a new PR with a proper implementation that won't break things?

**Comment by GdoongMathew** [comment]
> @Borda I see that you have reverted this PR and I see why: It seems like I inadvertently introduced [another bug](https://github.com/mlflow/mlflow/issues/15111), which I only noticed now since I am also affected :| Would you accept a new PR with a proper implementation that won't break things?

Hi @benglewis , do you think you still have the bandwidth to help with a workaround? as I think this could be a relatively simple fix -- just converting artifact `Path` object to a string before calling `self.experiment.log_artifact(self._run_id, p, artifact_path)`. It might also make sense to include this in the current master branch.
